docs/adr/000x-modelling-history.md

```diff
@@ -0,0 +1,314 @@
+
+Accepted
+
+## Consequences
+
+This modelling should provide a basis to provide the richer history
+functionality that can enable meeting and exceeding the features offered by
+Whitehall and Mainstream Publisher.
+
+It does however come with a number of concerns that are highlighted below.
+
+### Mo' models, mo' problems
+
+With an increase in the number of models there is more for a developer to
+understand and a higher barrier for entry. For every action there are either
+more database queries or a more complicated single query to look up information.
+
+Some ideas to alleviate this are:
+
+- the use of [delegate][] with ActiveRecord to make it simpler to access data
+  without interacting with individual models;
+- use of [scopes][] / find methods to alleviate join knowledge;
+- A [materialized view][] or [search index][elasticsearch-rails] to have
+  current and live edition data in a flat form.
+
+### We're not considering removing the past
+
+This data model is built around the expectation that the history of documents
+published to GOV.UK is something that we want to preserve. Over time this may
+lead to a large amount of data usage.
+
+## Further options being considered
+
+This section outlines some further options that was considered as part of this
+ADR but not part of the decision and may be explored further in future.
+
+### Efficient ways to load a timeline
+
+With the database storing a rich history there are options for the
+timeline to associate with models directly rather than store much information
+directly. However this may not be efficient or simple to access due to the
+need for many joins.
+
+A suggestion would be that we associate a TimelineEntry model with the
+associated models created in the event that created the entry but also store
+a denormalised version of the data tailored for rendering the timeline. If at
+any time we have to resolve an issue with incorrect entries or want to
+change their functionality we can use the associated model(s) as a source of
+truth to regenerate them.
+
+### States changed asynchronously
+
+This proposal mostly considers state changes to be a synchronous action in a
+single request. However this may be done asynchronously in order to provide
+timely responses and accurate information.
+
+This presents us with a challenge in that an edition can be in a transitory
+state and the way we present that might differ depending on the status. For
+example:
+
+- Transitioning from `draft` -> `submitted_for_review`. We'd consider the
+  document to instantly be in `submitted_for_review` status and no need to
+  update the Publishing API.
+- Transitioning from `scheduled` -> `published`. We'd consider the document to not
+  be considered published until we had successfully synced the resources with
+  the Publishing API. We'd want to inform the user that a transition was in
+  progress and whether that failed.
+
+An approach to resolve this would be to provide the ability for an edition to
```

Comment:
This is something that the design/product side are having to consider with regards to scheduling so the decision on that will be likely dictated from that.

My logic is that you're in the same status until the status change completes. E.g. if you're scheduled you're still there until publishing completes, if you're unpublishing you're published until that completes. These might not all hold though.

Pending kind of works unless the pending fails and then you end up in a state of published but not actually published. It could even be that there's a concept of a cover all "failed send help" status that would then allow a pending concept. 

Key outcome I'm hoping for with the scheduling work is that we're able to find a pattern that is suitable to scale to asynchronously allowing publishing/unpublishing/etc.

Reply:
Thanks for explaining more :-). I think our logic is quite similar; either way we end up with a piece of state - whether it's a pending flag on a status, or a separate pending status, that we need to do something with. Since we have to do work either way in case of failure/success, it seems to me like the flag approach is simpler, but perhaps I'm missing some advantage?

Reply:
No no, you're correct that a single status with a flag would be simpler.

It does feel a bit weird to me that you'd have something with a status of "published" that isn't actually published but if that fits into how product understands it then that'd be fine.

But yeah to me the key bit is what something is when a status transition fails, that approach there will probably dictate what option is applicable.

Btw I'm planning to drop this section from the ADR, but very happy to continue discussing our best options for this as this would be a really good problem to solve.